### ISSUE TITLE
Hello! :-)

### ISSUE THREAD

**Issue by floswald** [alancrawford/GEV.jl#1]
hey boss @alancrawford !

I've been looking for a clogit implementation and guess what - here we are! 

- status of this project?
- any thoughts about merging this into another package? like https://github.com/mattwigway/DiscreteChoiceModels.jl
- people are [looking for this](https://discourse.julialang.org/t/conditional-logit-clogit-in-julia/64215), so it's an important package.

**Comment by alancrawford** [comment]
Hi Flo!

I had no idea others were aware of this hobby project! 

Status - sleeping. It was written several years ago after despondency with stata's microdata nested logit estimation. At the time I had the idea of making the full suite of GEV models from Train's book in available in Julia with a usable interface. I also wanted to allow more than a handful of alternatives and be able to take advantage of AD and parallel processing. As you can see I didn't get that far in my wider goals, only as far as my paper needed. 

[DiscreteChoiceModels.jl ](https://github.com/mattwigway/DiscreteChoiceModels.jl) looks like a great initiative. I didn't know about it, but I am very supportive of it. 

Truth be told I don't have a ton of bandwidth to develop and merge this package and won't have going forward ... not sure what to suggest

**Comment by floswald** [comment]
well, came across it by chance just now. so let me get in touch with the other package author. just to make sure, what is there in terms of code works well for you (i.e. it's correct)? I can see any formal tests, so just asking ... ;-)

**Comment by alancrawford** [comment]
AFAIK it works fine for me. My "test" - such as it is - is the replication of stata output for matching commands... To the best of my recollection the output matches. If someone does want to grab bits of it and has questions, of course I am willing to help... I just don't have the resources to build a super reliable, maintained package.

**Comment by floswald** [comment]
sounds good to me. can I ask what was the main theoretical reference for you (and do you have a link to that)? thanks

**Comment by floswald** [comment]
yeah no worries, important thing is that it's correct, we can optimize performance and style later on. thanks!

**Comment by floswald** [comment]
man the data structure is so important in this stuff. also without docs 😰  
Im starting to understand though. also, the example runs! but only in it's current form - i wanted to wrap it into a function but that does not work because it expects to find `Main.cl.data` ie as a global variable. I'd like to get rid of that and supply `cl` as an argument to the functions (note that there is no data copying involved). any thoughts on that?

**Comment by alancrawford** [comment]
Yeah , the data structure is a bit of a pickle. I don't claim it is optimal and i was trying to use formulas to make it easier to work with. I found it worked OK for the user, but was not too elegant in the back end. 

On the global variable hack - now you are testing my memory. I think the reason for using a global variable was to avoid transportation of data across workers in every call of objective function and gradient. It only needs to be moved once. There is no reason in serial other than my being a bit lazy for my use case. 

I think I can remember trying to figure out the workerpool setup for pmap on machines not sharing memory. But i seem to recall being a little frustrated by the early Julia documentation (i think it was quite recently introduced when i was coding this) and was not able to get it to work as intended. So I made a filthy hack (which you have found in record time!) for my use case with global variables that sliced up the data and moved it just once. 

Not a good reason ... maybe in newer julia there are much better ways / better documentation / i always didn't get the key trick!

**Comment by floswald** [comment]
ok good, just to understand. thanks!
